### ISSUE TITLE
oci_network_firewall_network_firewall_policy

### ISSUE THREAD

**Issue by cristian-avramita** [oracle/terraform-provider-oci#1640]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version and Provider Version

```
terraform {
  required_version = ">= 1.0.0"
  required_providers {
    oci = {
      version = "4.85.0"
    }
  }
}
```

### Affected Resource(s)

```
oci_network_firewall_network_firewall_policy
```

### Expected Behavior

I want to create multiple ip_address_lists blocks for example:
```
    ip_address_lists {
    ip_address_list_name = "host1"
    ip_address_list_value = ["192.168.0.0/16"]
  }

    ip_address_lists {
    ip_address_list_name = "cidr1"
    ip_address_list_value = ["10.0.0.0/16"]
  }
```

So I should have host1 with "192.168.0.0/16" and cidr1 with "10.0.0.0/16".

### Actual Behavior

But as you can see below from tfstate the result is:
```
            "ip_address_lists": [
              {
                "ip_address_list_name": "host1",
                "ip_address_list_value": [
                  "192.168.0.0/16"
                ]
              },
              {
                "ip_address_list_name": "host1",
                "ip_address_list_value": [
                  "192.168.0.0/16"
                ]
              }
            ],
```

The same strange behavior happens for application_lists:
```
            "application_lists": [
              {
                "icmp_code": 0,
                "icmp_type": 0,
                "key": "http",
                "maximum_port": 22,
                "minimum_port": 22,
                "type": "TCP"
              },
              {
                "icmp_code": 0,
                "icmp_type": 0,
                "key": "http",
                "maximum_port": 80,
                "minimum_port": 80,
                "type": "TCP"
              }
            ],
```

And also for url_lists:
```
            "url_lists": [
              {
                "key": "google",
                "pattern": "youtube.com",
                "type": "SIMPLE"
              },
              {
                "key": "google",
                "pattern": "google.com",
                "type": "SIMPLE"
              }
            ]
```

I have also created manually in Oracle Cloud the network firewall policy and after that I have downloaded the generated stack.
And for example the output for ip_address_lists is the following:
```
  ip_address_lists {
    ip_address_list_name = "host1"
    ip_address_list_value = [
      "192.168.88.100",
    ]
  }
  ip_address_lists {
    ip_address_list_name = "host1"
    ip_address_list_value = [
      "192.168.88.100",
    ]
  }
```
Again no cidr1 key found in ip_address_lists.

Another issue that I seen is for "decryption_rules".
When I created one rule with 
action = "NO_DECRYPT" 
decryption_profile = null
I received the following error
```
Error: 400-CannotParseRequest, Incorrectly formatted request. decryption-profile should not be present in decryption-rule decr2 with action no decrypt Please refer to our documentation for help.
```

I don't want to entirely delete the parameter "decryption_profile" because I will create multiple rules using a dynamic block with for_each (some will be NO_DECRYPT and other will be DECRYPT).

PS: Even if I have created a separated block (for action = "NO_DECRYPT"), that doesn't contain "decryption_profile " and "secret" I receive the same error.

Steps to reproduce:
- Run "terraform apply" in order to create oci_network_firewall_network_firewall_policy with one decryption_rules as below example (no decryption_profile or secret is present in decryption_rules ):
```
  dynamic "decryption_rules" {
    for_each = each.value.no_decryption_rules
    iterator = rr
    content {
      name   = rr.value.name
      action = rr.value.action
      condition {
        destinations = rr.value.destinations
        sources      = rr.value.sources
      }
    }
  }
```

- After the resource is created change a parameter (ex ip_address_list_value) and run again "terraform apply" and the result is:
```
│ Error: 400-CannotParseRequest, Incorrectly formatted request. decryption-profile should not be present in decryption-rule decr1 with action no decrypt Please refer to our documentation for help. 
│ Suggestion: Please retry or contact support for help with service: Network Firewall Policy
│ Documentation: https://registry.terraform.io/providers/oracle/oci/latest/docs/resources/network_firewall_network_firewall_policy 
│ API Reference:  
│ Request Target: PUT https://network-firewall.eu-amsterdam-1.ocs.oraclecloud.com/20211001/networkFirewallPolicies/xxxxxxxxx 
│ Provider version: 4.85.0, released on 2022-07-25.  
│ Service: Network Firewall Policy 
│ Operation Name: UpdateNetworkFirewallPolicy 
│ OPC request ID: xxxxxxxxx 
```

- Run 
```
terraform taint module.ams-network-firewall.oci_network_firewall_network_firewall_policy.this[\"ams_firewall_policy\"]
terraform apply
```

- The result is
```
module.ams-network-firewall.oci_network_firewall_network_firewall_policy.this["ams_firewall_policy"]: Creation complete after 0s [id=xxxxxxx]

Apply complete! Resources: 1 added, 0 changed, 1 destroyed.
```

**Comment by jakemccoynetworkguy** [comment]
I have hit this bug :/ easily replicated, the provider seems to combine the two application_lists into one

```
resource "oci_network_firewall_network_firewall_policy" "firewall_policy" {
..

  application_lists {
    key          = "App_List_1"
    minimum_port = "22"
    maximum_port = "22"
    type         = "TCP"
  }
  application_lists {
    key          = "App_List_2"
    minimum_port = "25"
    maximum_port = "25"
    type         = "TCP"
  }
```
Then the provider combines into one (seemly the last one, like the key was reused from the last iteration vs the current iteration)
Remote state shows the result 
```
            "application_lists": [
              {
                "icmp_code": 0,
                "icmp_type": 0,
                "key": "App_List_2",
                "maximum_port": 22,
                "minimum_port": 22,
                "type": "TCP"
              },
              {
                "icmp_code": 0,
                "icmp_type": 0,
                "key": "App_List_2",
                "maximum_port": 25,
                "minimum_port": 25,
                "type": "TCP"
              }
            ],
            ```

**Comment by ravinitp** [comment]
Thank you for reporting the issue. We have raised an internal ticket to track this. Our service engineers will get back to you.

**Comment by ravinitp** [comment]
The fix is  released and available  in version 4.112.0 and above (Released on 15/Mar/23). Please use latest version.
